services/construction_service.go

```diff
@@ -365,6 +369,54 @@ func (s *ConstructionAPIService) ConstructionMetadata(
 		return nil, wrapErr(ErrGeth, err)
 	}
 
+	// Build eth transaction for L1 fee calculation
+	var toAddress common.Address
+	var ethTx *ethTypes.Transaction
+	if to != "" {
+		toAddress = common.HexToAddress(to)
+	}
+	if isEIP1559 := gasTipCap != nil && gasFeeCap != nil; isEIP1559 {
+		ethTx = ethTypes.NewTx(&ethTypes.DynamicFeeTx{
+			Nonce:     nonce,
+			GasTipCap: gasTipCap,
+			GasFeeCap: gasFeeCap,
+			Gas:       gasLimit,
+			To:        &toAddress,
+			Value:     input.Value,
+			Data:      input.Data,
+		})
+	} else {
+		ethTx = ethTypes.NewTx(&ethTypes.LegacyTx{
+			Nonce:    nonce,
+			GasPrice: gasPrice,
+			Gas:      gasLimit,
+			To:       &toAddress,
+			Value:    input.Value,
+			Data:     input.Data,
+		})
+	}
+	ethTxBytes, err := ethTx.MarshalBinary()
+	if err != nil {
+		return nil, wrapErr(ErrL1DataFee, err)
+	}
+
+	// Get L1 data fee
+	var l1Fee *big.Int
+	if s.config.GethURL != "" {
+		client, err := ethclient.DialContext(ctx, s.config.GethURL)
```

Comment:
This creates a TCP connection for every /metadata call. This is slow and expensive.
Better to reuse the existing client connection in the `ConstructionAPIService`. The bindings expect a type that implements the [ContractBackend](https://github.com/ethereum-optimism/op-geth/blob/cd4a1e46153da15e37a3fa486b9ee4e08efa4734/accounts/abi/bind/backend.go#L122) interface. The existing [services.Client](https://github.com/Inphi/optimism-rosetta/blob/master/services/types.go#L31) interface already partially satisfies `ContractBackend`. Just need to add and implement a couple more functions.

Reply:
good point, let me check how many functions are left from `services.Client`

Reply:
done